[CALCITE-4636] Switch out of agg mode when constructing RelCollation …

a-rafay opened the pull request:
…for aggregate functions (Rafay)

zinking commented:
LGTM. does it only generates impact on your hypothetical query? or other queries are wrong as well ?

a-rafay commented:
> LGTM. does it only generates impact on your hypothetical query? or other queries are wrong as well ?

@zinking the test case i fixed in this patch was wrong. So there might be other queries giving wrong results. I found one in my test case too, but it was very similar to the test case which i fixed.

amaliujia commented on core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:
```diff
@@ -5758,14 +5758,21 @@ private void translateAgg(SqlCall call, @Nullable SqlNode filter,
-                .collect(Collectors.toList()));
+        try {
+          // switch out of agg mode
+          bb.agg = null;
```
My question is why not keep the reference of bb.agg and then in the finally block, assign that reference back? (or is the bb.agg is always equal to this class?)

a-rafay replied:
It must be equal to this class. If you scroll at the top of this function, you'll see that. We also do something similar earlier in this function as well.

a-rafay replied:
See: https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java#L5585
and:
https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java#L5700

amaliujia replied:
Thanks for clarifying. 

I can see now this is already an existing practice.

a-rafay commented:
@amaliujia any update on this?

amaliujia commented:
Fixed via https://github.com/apache/calcite/commit/4e679f03771c89ffa463051b5698ead1d935748b
